File: src/lib/features/engine.ts
// Feature Engine - Orchestrates all feature calculations

import { PriceData, preparePriceSeries } from "./calculator";
import { calculateTrendScore, TrendResult } from "./trend";
import { calculateDerivativeScore, DerivativeResult } from "./derivative";
import { calculateVolumeScore, VolumeResult } from "./volume";
import { calculateMomentumScore, MomentumResult } from "./momentum";
import { calculateConfidence, ConfidenceResult, ConfidenceWeights } from "./confidence";

export interface FeatureEngineResult {
  trend_score: number;
  derivative_score: number;
  volume_score: number;
  momentum_score: number;
  trend_detail: TrendResult["detail"];
  derivative_detail: DerivativeResult["detail"];
  volume_detail: VolumeResult["detail"];
  momentum_detail: MomentumResult["detail"];
  confidence_score: number;
  data_completeness: number;
  missing_sources: string[];
  error?: string;
}

export interface SourceOk {
  binance_spot: boolean;
  binance_futures: boolean;
  coingecko: boolean;
}

/**
 * Run full feature pipeline for a single coin
 */
export function runFeatureEngine(
  priceData: PriceData[],
  metrics: {
    openInterest: number | null;
    openInterestPrev?: number | null;
    fundingRate: number | null;
    marketCap: number | null;
  },
  healthWeights: any,
  confidenceWeights: ConfidenceWeights,
  sourceOk?: SourceOk
): FeatureEngineResult {
  // Validate minimum data
  if (priceData.length < 20) {
    return {
      trend_score: 50,
      derivative_score: 50,
      volume_score: 50,
      momentum_score: 50,
      trend_detail: {
        price: 0,
        ema20: 0,
        ema50: 0,
        ema200: 0,
        price_vs_ema20: false,
        price_vs_ema50: false,
        price_vs_ema200: false,
        ema20_vs_ema50: false,
        ema50_vs_ema200: false,
        score_breakdown: { base: 50 },
      },
      derivative_detail: {
        oi_current: null,
        oi_prev: null,
        oi_change_pct: 0,
        funding_rate: null,
        oi_component: 50,
        funding_component: 50,
        accumulation_bonus: 0,
        no_futures: true,
      },
      volume_detail: {
        volume_current: 0,
        volume_ma20: 0,
        volume_ratio: 1,
        days_used: 0,
      },
      momentum_detail: {
        roc_14: 0,
        atr_14: 0,
        atr_pct: 0,
        roc_component: 50,
        atr_component: 50,
      },
      confidence_score: 0,
      data_completeness: 0,
      missing_sources: ["binance_spot", "binance_futures", "coingecko"],
      error: "Insufficient price data (need >= 20 rows)",
    };
  }

  const { closes, highs, lows, volumes } = preparePriceSeries(priceData);

  const trendResult = calculateTrendScore(closes);
  const volumeResult = calculateVolumeScore(volumes);
  const momentumResult = calculateMomentumScore(closes, highs, lows);
  const hasFutures = metrics.openInterest !== null || metrics.fundingRate !== null;
  const derivativeResult = calculateDerivativeScore(
    metrics.openInterest,
    metrics.openInterestPrev ?? null,
    metrics.fundingRate,
    hasFutures
  );

  const confidenceResult = calculateConfidence(
    sourceOk?.binance_spot || false,
    sourceOk?.binance_futures || false,
    sourceOk?.coingecko || false,
    hasFutures,
    confidenceWeights
  );

  return {
    trend_score: trendResult.score,
    derivative_score: derivativeResult.score,
    volume_score: volumeResult.score,
    momentum_score: momentumResult.score,
    trend_detail: trendResult.detail,
    derivative_detail: derivativeResult.detail,
    volume_detail: volumeResult.detail,
    momentum_detail: momentumResult.detail,
    confidence_score: confidenceResult.confidence_score,
    data_completeness: confidenceResult.data_completeness,
    missing_sources: confidenceResult.missing_sources,
  };
}

/**
 * Calculate health score from feature scores
 */
export function calculateHealthScore(
  trendScore: number,
  derivativeScore: number,
  volumeScore: number,
  momentumScore: number,
  weights: { trend: number; derivative: number; volume: number; momentum: number } = {
    trend: 0.35,
    derivative: 0.35,
    volume: 0.2,
    momentum: 0.1,
  }
): number {
  const score =
    trendScore * weights.trend +
    derivativeScore * weights.derivative +
    volumeScore * weights.volume +
    momentumScore * weights.momentum;

  return Number(Math.max(0, Math.min(100, score)).toFixed(1));
}

/**
 * Get recommendation signal from health score
 */
export function getRecommendationSignal(
  healthScore: number,
  thresholds: { strong_watch: number; watch: number; observe: number } = {
    strong_watch: 90,
    watch: 80,
    observe: 65,
  }
): "STRONG_WATCH" | "WATCH" | "OBSERVE" | "WEAK" {
  if (healthScore >= thresholds.strong_watch) return "STRONG_WATCH";
  if (healthScore >= thresholds.watch) return "WATCH";
  if (healthScore >= thresholds.observe) return "OBSERVE";
  return "WEAK";
}

/**
 * Generate recommendation reason text
 */
export function generateRecommendationReason(
  signal: string,
  trendScore: number,
  derivativeScore: number,
  volumeScore: number,
  momentumScore: number,
  confidenceScore: number
): string {
  const parts: string[] = [];

  // Signal-based opening
  switch (signal) {
    case "STRONG_WATCH":
      parts.push("Strong bullish signals across all metrics.");
      break;
    case "WATCH":
      parts.push("Positive indicators with room for monitoring.");
      break;
    case "OBSERVE":
      parts.push("Mixed signals, continue observing.");
      break;
    case "WEAK":
      parts.push("Weak signals, exercise caution.");
      break;
  }

  // Add specific insights
  if (trendScore >= 75) {
    parts.push("Price above key EMAs.");
  } else if (trendScore < 40) {
    parts.push("Price below key EMAs.");
  }

  if (derivativeScore >= 75) {
    parts.push("Derivatives show accumulation.");
  } else if (derivativeScore < 40) {
    parts.push("Derivatives show distribution.");
  }

  if (volumeScore >= 75) {
    parts.push("Volume significantly above average.");
  } else if (volumeScore < 40) {
    parts.push("Volume below average.");
  }

  if (momentumScore >= 75) {
    parts.push("Strong momentum.");
  } else if (momentumScore < 40) {
    parts.push("Weak momentum.");
  }

  // Confidence warning
  if (confidenceScore < 70) {
    parts.push(`⚠ Data confidence: ${confidenceScore.toFixed(0)}%`);
  }

  return parts.join(" ");
}

File: backend/features/engine.py
"""
Feature Engine - Orchestrates all feature calculations
"""
import pandas as pd
from typing import Dict, Any, Optional, List

from backend.features.trend import calculate_trend_score
from backend.features.derivative import calculate_derivative_score
from backend.features.volume import calculate_volume_score
from backend.features.momentum import calculate_momentum_score
from backend.features.confidence import calculate_confidence


class FeatureEngine:
    """Orchestrates feature calculation for coins"""

    def __init__(
        self,
        health_weights: Dict[str, float] = None,
        confidence_weights: Dict[str, float] = None,
        recommendation_thresholds: Dict[str, int] = None,
    ):
        self.health_weights = health_weights or {
            "trend": 0.35,
            "derivative": 0.35,
            "volume": 0.20,
            "momentum": 0.10,
        }
        self.confidence_weights = confidence_weights or {
            "binance_spot": 0.30,
            "binance_futures": 0.40,
            "coingecko": 0.30,
        }
        self.recommendation_thresholds = recommendation_thresholds or {
            "strong_watch": 90,
            "watch": 80,
            "observe": 65,
        }

    def run(
        self,
        price_data: List[Dict[str, Any]],
        oi_current: Optional[float],
        oi_prev: Optional[float],
        funding_rate: Optional[float],
        has_futures: bool,
        source_ok: Dict[str, bool],
    ) -> Dict[str, Any]:
        """
        Run full feature pipeline for a single coin
        
        price_data: List of {date, open, high, low, close, volume}
        """
        # Convert to DataFrame
        if not price_data:
            return self._empty_result(has_futures)

        df = pd.DataFrame(price_data)

        if len(df) < 20:
            return self._empty_result(has_futures, error="Insufficient price data (need >= 20 rows)")

        # Calculate features
        trend_result = calculate_trend_score(df)
        volume_result = calculate_volume_score(df)
        momentum_result = calculate_momentum_score(df)
        derivative_result = calculate_derivative_score(
            oi_current, oi_prev, funding_rate, has_futures
        )
        confidence_result = calculate_confidence(
            binance_spot_ok=source_ok.get("binance_spot", False),
            binance_futures_ok=source_ok.get("binance_futures", False),
            coingecko_ok=source_ok.get("coingecko", False),
            has_futures=has_futures,
            weights=self.confidence_weights,
        )

        return {
            "trend_score": trend_result["score"],
            "derivative_score": derivative_result["score"],
            "volume_score": volume_result["score"],
            "momentum_score": momentum_result["score"],
            "trend_detail": trend_result["detail"],
            "derivative_detail": derivative_result["detail"],
            "volume_detail": volume_result["detail"],
            "momentum_detail": momentum_result["detail"],
            "confidence_score": confidence_result["confidence_score"],
            "data_completeness": confidence_result["data_completeness"],
            "missing_sources": confidence_result["missing_sources"],
        }

    def calculate_health_score(
        self,
        trend_score: float,
        derivative_score: float,
        volume_score: float,
        momentum_score: float,
    ) -> float:
        """Calculate weighted health score"""
        score = (
            trend_score * self.health_weights["trend"] +
            derivative_score * self.health_weights["derivative"] +
            volume_score * self.health_weights["volume"] +
            momentum_score * self.health_weights["momentum"]
        )
        return round(max(0, min(100, score)), 1)

    def get_health_status(self, score: float) -> str:
        """Get health status from score"""
        if score >= 90:
            return "STRONG"
        if score >= 80:
            return "HEALTHY"
        if score >= 65:
            return "NEUTRAL"
        if score >= 50:
            return "CAUTION"
        return "WEAK"

    def get_recommendation_signal(self, health_score: float) -> str:
        """Get recommendation signal from health score"""
        if health_score >= self.recommendation_thresholds["strong_watch"]:
            return "STRONG_WATCH"
        if health_score >= self.recommendation_thresholds["watch"]:
            return "WATCH"
        if health_score >= self.recommendation_thresholds["observe"]:
            return "OBSERVE"
        return "WEAK"

    def generate_recommendation_reason(
        self,
        signal: str,
        trend_score: float,
        derivative_score: float,
        volume_score: float,
        momentum_score: float,
        confidence_score: float,
    ) -> str:
        """Generate recommendation reason text"""
        parts = []

        # Signal-based opening
        if signal == "STRONG_WATCH":
            parts.append("Strong bullish signals across all metrics.")
        elif signal == "WATCH":
            parts.append("Positive indicators with room for monitoring.")
        elif signal == "OBSERVE":
            parts.append("Mixed signals, continue observing.")
        else:
            parts.append("Weak signals, exercise caution.")

        # Add specific insights
        if trend_score >= 75:
            parts.append("Price above key EMAs.")
        elif trend_score < 40:
            parts.append("Price below key EMAs.")

        if derivative_score >= 75:
            parts.append("Derivatives show accumulation.")
        elif derivative_score < 40:
            parts.append("Derivatives show distribution.")

        if volume_score >= 75:
            parts.append("Volume significantly above average.")
        elif volume_score < 40:
            parts.append("Volume below average.")

        if momentum_score >= 75:
            parts.append("Strong momentum.")
        elif momentum_score < 40:
            parts.append("Weak momentum.")

        # Confidence warning
        if confidence_score < 70:
            parts.append(f"⚠ Data confidence: {confidence_score:.0f}%")

        return " ".join(parts)

    def _empty_result(self, has_futures: bool, error: str = None) -> Dict[str, Any]:
        return {
            "trend_score": 50.0,
            "derivative_score": 50.0,
            "volume_score": 50.0,
            "momentum_score": 50.0,
            "trend_detail": {},
            "derivative_detail": {"no_futures": not has_futures},
            "volume_detail": {},
            "momentum_detail": {},
            "confidence_score": 0.0,
            "data_completeness": 0.0,
            "missing_sources": ["binance_spot", "binance_futures", "coingecko"],
            "error": error,
        }
